src/com/google/javascript/jscomp/deps/ModuleLoader.java

```diff
@@ -388,51 +424,55 @@ public static boolean isPathIdentifier(String name) {
-      builder.add(ModuleNames.escapePath(resolver.apply(root)));
+      builder.add(new InputPathData(ModuleNames.escapePath(resolver.apply(root))));
     }
     return builder.build();
   }
 
-  private static ImmutableSet<String> resolvePaths(
-      Iterable<String> names, ImmutableList<String> roots) {
-    HashSet<String> resolved = new HashSet<>();
+  private static ImmutableMap<String, InputPathData> resolvePaths(
+      Iterable<String> names, ImmutableList<InputPathData> roots) {
+    Map<String, InputPathData> resolved = new HashMap<>();
+    Set<String> knownPaths = new HashSet<String>();
     for (String name : names) {
-      if (!resolved.add(normalize(ModuleNames.escapePath(name), roots))) {
+      InputPathData inputPathData = new InputPathData(ModuleNames.escapePath(name));
+      InputPathData normalizedPathData = normalize(inputPathData.path, roots);
+      if (!knownPaths.add(normalizedPathData.path)) {
         // Having root paths "a" and "b" and source files "a/f.js" and "b/f.js" is ambiguous.
         throw new IllegalArgumentException(
             "Duplicate module path after resolving: " + name);
       }
+      resolved.put(inputPathData.ensureLeadingSlash(), inputPathData);
     }
-    return ImmutableSet.copyOf(resolved);
+    return ImmutableMap.copyOf(resolved);
   }
 
-  /**
-   * Normalizes the name and resolves it against the module roots.
-   */
-  private static String normalize(String path, Iterable<String> moduleRootPaths) {
+  /** Normalizes the name and resolves it against the module roots. */
+  private static InputPathData normalize(String path, Iterable<InputPathData> moduleRootPaths) {
     // Find a moduleRoot that this URI is under. If none, use as is.
-    for (String moduleRoot : moduleRootPaths) {
-      if (path.startsWith(moduleRoot)) {
+    InputPathData inputPath = new InputPathData(path);
+    for (InputPathData moduleRoot : moduleRootPaths) {
+      if (inputPath.ensureLeadingSlash().startsWith(moduleRoot.ensureLeadingSlash())) {
         // Make sure that e.g. path "foobar/test.js" is not matched by module "foo", by checking for
         // a leading slash.
-        String trailing = path.substring(moduleRoot.length());
+        String trailing =
+            inputPath.ensureLeadingSlash().substring(moduleRoot.ensureLeadingSlash().length());
         if (trailing.startsWith(MODULE_SLASH)) {
-          return trailing.substring(MODULE_SLASH.length());
+          return new InputPathData(trailing.substring(MODULE_SLASH.length()));
         }
       }
     }
     // Not underneath any of the roots.
-    return path;
+    return inputPath;
   }
 
   /** Build the module registry from the set of module paths */
-  private static ImmutableSortedMap<String, ImmutableSet<String>> buildRegistry(
-      ImmutableSet<String> modulePaths) {
-    SortedMap<String, Set<String>> registry =
+  private static ImmutableSortedMap<String, ImmutableMap<String, InputPathData>> buildRegistry(
```

Comment:
Please expand the JavaDoc to define the meaning of the keys and values of both the input and output.